File: Fall Week 4 Binary Search/python/solutions/B.py
'''
Using a prefix sum (the sum of the first 'i' elements for all 'i')
we can quickly check whether the sum of the first 'i' piles is biggr
than the query!
'''

n = int(input())
a = list(map(int, input().split()))
m = int(input())
q = list(map(int, input().split()))

prefix = [0] * n
prefix[0] = a[0]
for i in range(1, n):
    prefix[i] = prefix[i - 1] + a[i]

for i in range(m):
    lo, hi = 0, n - 1
    while hi - lo > 3:
        mid = (lo + hi) // 2
        if prefix[mid] < q[i]:
            lo = mid
        else:
            hi = mid


    for j in range(lo, hi + 1):
        if prefix[j] >= q[i]:
            print(j + 1)
            break
        

File: Fall Week 4 Binary Search/c++/solutions/B.cpp
#include "bits/stdc++.h"
using namespace std;

/*
Using a prefix sum (the sum of the first 'i' elements for all 'i')
we can quickly check whether the sum of the first 'i' piles is biggr
than the query!
 */

int main(){
    int n;
    cin >> n;
    vector<int> a(n);
    for(int i = 0; i < n; i++){
        cin >> a[i];
    }
    int m;
    cin >> m;
    vector<int> q(m);
    for(int i = 0; i < m; i++){
        cin >> q[i];
    }

    vector<int> prefix(n);
    prefix[0] = a[0];
    for(int i = 1; i < n; i++){
        prefix[i] = prefix[i - 1] + a[i];
    }

    for(int i : q){
        int lo = 0, hi = n - 1;
        while(hi - lo > 3){
            int mid = (lo + hi) / 2;
            if(prefix[mid] < i){
                lo = mid;
            } else {
                hi = mid;
            }
        }
        for(int j = lo; j <= hi; j++ ){
            if(prefix[j] >= i){
                cout << j + 1 << endl;
                break;
            }
        }
    }
}
